src/main/java/org/mastodon/mamut/feature/branch/divisions/BranchCellDivisionFrequencyFeatureComputer.java

```diff
@@ -28,65 +28,88 @@
  */
 package org.mastodon.mamut.feature.branch.divisions;
 
-import org.mastodon.mamut.feature.AbstractSerialFeatureComputer;
+import net.imglib2.util.Pair;
+import net.imglib2.util.ValuePair;
+import org.mastodon.mamut.feature.AbstractResettableFeatureComputer;
 import org.mastodon.mamut.feature.MamutFeatureComputer;
-import org.mastodon.mamut.feature.ValueIsSetEvaluator;
 import org.mastodon.mamut.feature.branch.BranchSpotFeatureUtils;
+import org.mastodon.mamut.model.branch.BranchLink;
 import org.mastodon.mamut.model.branch.BranchSpot;
+import org.mastodon.mamut.model.branch.ModelBranchGraph;
+import org.mastodon.mamut.util.LineageTreeUtils;
 import org.mastodon.properties.DoublePropertyMap;
+import org.mastodon.properties.ObjPropertyMap;
 import org.scijava.ItemIO;
 import org.scijava.plugin.Parameter;
 import org.scijava.plugin.Plugin;
 
-import java.util.Collection;
+import javax.annotation.Nonnull;
 
 /**
  * Computes {@link BranchCellDivisionFrequencyFeature}
  */
 @Plugin( type = MamutFeatureComputer.class )
-public class BranchCellDivisionFrequencyFeatureComputer extends AbstractSerialFeatureComputer< BranchSpot >
+public class BranchCellDivisionFrequencyFeatureComputer extends AbstractResettableFeatureComputer
 {
 
+	@Parameter
+	protected ModelBranchGraph branchGraph;
+
 	@Parameter( type = ItemIO.OUTPUT )
 	protected BranchCellDivisionFrequencyFeature output;
 
-	@Override
-	protected void compute( final BranchSpot branchSpot )
-	{
-		output.nCellDivisions.set( branchSpot, cellDivisionFrequency( branchSpot ) );
-	}
+	private ObjPropertyMap< BranchSpot, Pair< Integer, Integer > > valueCache;
 
 	@Override
 	public void createOutput()
 	{
 		if ( null == output )
-			output = new BranchCellDivisionFrequencyFeature(
-					new DoublePropertyMap<>( model.getBranchGraph().vertices().getRefPool(), Double.NaN ) );
-	}
-
-	private double cellDivisionFrequency( final BranchSpot branchSpot )
-	{
-		int leaves = BranchSpotFeatureUtils.countLeaves( model.getBranchGraph(), branchSpot );
-		int divisions = leaves - 1; // NB: we are assuming a binary tree
-		int totalDuration = BranchSpotFeatureUtils.totalBranchDurations( model.getBranchGraph(), branchSpot );
-		return divisions / ( double ) totalDuration;
+			output = new BranchCellDivisionFrequencyFeature( new DoublePropertyMap<>( branchGraph.vertices().getRefPool(), Double.NaN ) );
 	}
 
 	@Override
-	protected void reset()
+	public void run()
 	{
-		output.nCellDivisions.beforeClearPool();
+		super.run();
+		valueCache = new ObjPropertyMap<>( branchGraph.vertices().getRefPool(), 0 );
+		LineageTreeUtils.callDepthFirst( branchGraph, this::computeCellDivisionFrequency, this::isCanceled );
 	}
 
-	@Override
-	protected ValueIsSetEvaluator< BranchSpot > getEvaluator()
+	private void computeCellDivisionFrequency( @Nonnull BranchSpot branchSpot )
 	{
-		return output;
+		if ( output.valueIsSet( branchSpot ) && !forceComputeAll.get() )
```

Comment:
The idea to recompute the feature only for a subset of branchspots is complicated. Whenever a branchspot changes, all the values for the predecessors would need to be updated too. So actually the invalidation of feature values would need to be corrected. But this is not relevant in practice for the current branchgraph implementation. Since it is never partially changed, it is only complete rebuilt.

Reply:
Thanks for this reminder. I downgraded this FeatureComputer from a **resettable** feature computer (where `force compute all` can be applied to) to only a **cancelable** feature computer, which only allows to cancel it during computation.